prescient.el

```diff
@@ -714,8 +717,277 @@ Return the sorted list. The original is modified destructively."
+the function `prescient-sort' (which does not depend on filtering
+input), then you might not need this option.
 
+See also the user option `prescient-sort-length-enable'."
+  :type 'boolean)
+
+(defface prescient-primary-highlight
+  '((t :weight bold))
+  "Face used to highlight the parts of candidates that match the input.")
+
+(defface prescient-secondary-highlight
+  '((t :inherit prescient-primary-highlight :underline t))
+  "Additional face used to highlight parts of candidates.
+
+May be used to highlight parts of candidates that match specific
+parts of the input.")
+
+;;;;; Utility functions
+
+;; From Orderless
+;; TODO: Not sure what this is for.
+(defun prescient--anchored-quoted-regexp (regexp)
+  "Determine if REGEXP is a quoted regexp anchored at the beginning.
+If REGEXP is of the form \"\\(?:^q\\)\" for q = (regexp-quote u),
+then return (cons REGEXP u); else return nil."
+  (when (and (string-prefix-p "\\(?:^" regexp) (string-suffix-p "\\)" regexp))
+    (let ((trimmed (substring regexp 5 -2)))
+      (unless (string-match-p "[$*+.?[\\^]"
+                              (replace-regexp-in-string
+                               "\\\\[$*+.?[\\^]" "" trimmed
+                               'fixedcase 'literal))
+        (cons regexp
+              (replace-regexp-in-string "\\\\\\([$*+.?[\\^]\\)" "\\1"
+                                        trimmed 'fixedcase))))))
+
+(defun prescient--highlight-completions (input candidates)
+  "According to INPUT, return list of propertized CANDIDATES."
+  (let ((regexps (prescient-filter-regexps input 'with-group))
+        (case-fold-search (if (eq prescient-use-case-folding
+                                  'smart)
+                              (let ((case-fold-search nil))
+                                ;; If using upper-case characters,
+                                ;; then don't fold case.
+                                (not (string-match-p "[[:upper:]]"
+                                                     input)))
+                            prescient-use-case-folding)))
+    (save-match-data
+      (mapcar
+       (lambda (candidate)
+         (setq candidate (copy-sequence candidate))
+         (prog1 candidate
+           (dolist (regexp regexps)
+             (when (string-match regexp candidate)
+               (font-lock-prepend-text-property
+                (match-beginning 0) (match-end 0)
+                'face 'prescient-primary-highlight
+                candidate)
+               (cl-loop
+                for (start end)
+                on (cddr (match-data))
+                by #'cddr
+                do (when (and start end)
+                     (font-lock-prepend-text-property
+                      start end
+                      'face 'prescient-secondary-highlight
+                      candidate)))))))
+       candidates))))
+
+;;;;; Sorting functions
+
+(defun prescient--sort-full-matches-first (candidates regexps ignore-case)
```

Comment:
Some part of this function has been copied from prescient-filter. I recommend to not copy code. Instead decompose prescient-filter in smaller functions and reuse them here. See also my comment about prescient--completion-style-filter, which is redundant with prescient-filter. You could refactor such that these two functions are unified as one.

Reply:
Generally - refactor first in preparation of the new feature and then add the new feature. This will help you in avoiding code duplication.